### PR TITLE
Psionic "Heal Other" Powers

### PR DESCRIPTION
**Pull request by VMSolidus**
# Description

This PR introduces two new Psionic Powers, Healing Word, and Breath of Life, both utilizing a new PsionicHealOtherSystem, which operates on datafield event arguments rather than a "hardcoded" component. Thus, any number of powers can be created which share this system. 

Healing Word is a power that features a short cast time, and heals a small amount of each damage type to a target(while reducing the target's rot timer slightly). It has a relatively short cooldown, and a low glimmer cost. 

Breath of Life by contrast, is an extremely rare power with a longer cast time, healing a much larger amount of each damage type to a target, reduces rot significantly, and attempts to revive the target. It has a 2 minute cooldown, and a high glimmer cost.

<details><summary><h1>Media</h1></summary>
<p>

image

</p>
</details>

# Changelog

:cl:
- add: Healing Word has been added as a new Psionic Power. When cast on another person, it heals a small amount of every damage type(scaling with Casting Stats), while also reducing rot timers. Healing Word has a very short cooldown, and a fairly low Glimmer cost.
- add: Breath of Life has been added as a new extremely rare Psionic Power. When cast on another person, it heals a large amount of damage(scaling with Casting Stats), while also substantially reducing rot timers. Additionally, it will revive the target if it is possible to do so. Breath of Life has an incredibly long cooldown, a long interuptable cast time, and an extraordinarily high glimmer cost(A typical Psion will spike glimmer by more than 50 points when casting it).
- add: The Chaplain now starts with the Healing Word power.

### CODE REVIEW

**Review by OldDanceJacket — APPROVED** [review]

File: Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    [Dependency] private readonly ExamineSystemShared _examine = default!;
+    [Dependency] private readonly DamageableSystem _damageable = default!;
+    [Dependency] private readonly RottingSystem _rotting = default!;
+    [Dependency] private readonly MobThresholdSystem _mobThreshold = default!;
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly GlimmerSystem _glimmer = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs:37]:
```suggestion
        base.Initialize();

```

File: Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        SubscribeLocalEvent<PsionicComponent, PsionicHealOtherPowerActionEvent>(OnPowerUsed);
+        SubscribeLocalEvent<PsionicComponent, DispelledEvent>(OnDispelled);
+        SubscribeLocalEvent<PsionicComponent, PsionicHealOtherDoAfterEvent>(OnDoAfter);
+    }
+
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs:42]:
```suggestion


```

File: Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+    [Dependency] private readonly SharedPsionicAbilitiesSystem _psionics = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly ExamineSystemShared _examine = default!;
+    [Dependency] private readonly DamageableSystem _damageable = default!;
+    [Dependency] private readonly RottingSystem _rotting = default!;
+    [Dependency] private readonly MobThresholdSystem _mobThreshold = default!;
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly GlimmerSystem _glimmer = default!;
+
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs:34]:
```suggestion


```

File: Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        _doAfterSystem.Cancel(component.DoAfter);
+        component.DoAfter = null;
+        args.Handled = true;
+    }
+
+    private void OnDoAfter(EntityUid uid, PsionicComponent component, PsionicHealOtherDoAfterEvent args)
+    {
+        if (component is null)
+            return; // It's entirely possible for the caster to stop being Psionic(due to mindbreaking) mid cast.
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs:104]:
```suggestion
        // It's entirely possible for the caster to stop being Psionic(due to mindbreaking) mid cast
        if (component is null)
            return;
```

File: Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        component.DoAfter = null;
+
+        if (args.Target is null)
+            return; // The target can also cease existing mid-cast.
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Abilities/Psionics/Abilities/HealOtherPowerSystem.cs:108]:
```suggestion
        // The target can also cease existing mid-cast
        if (args.Target is null)
            return;
```

File: Content.Shared/Actions/Events/PsionicHealOtherPowerActionEvent.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+{
+    [DataField]
+    public DamageSpecifier HealingAmount = default!;
+
+    [DataField]
+    public string PowerName;
+
+    /// <summary>
+    ///     Controls whether or not a power fires immediately and with no DoAfter.
+    /// </summary>
```
Comment (DEATHB4DEFEAT) [inline Content.Shared/Actions/Events/PsionicHealOtherPowerActionEvent.cs:16]:
```suggestion
    /// Controls whether or not a power fires immediately and with no DoAfter
```

File: Content.Shared/Psionics/Events.cs [hunk 7]
```diff
@@ -1,28 +1,59 @@
...
+    [DataField]
+    public bool DoRevive;
+
+    public PsionicHealOtherDoAfterEvent(TimeSpan startedAt)
+    {
+        StartedAt = startedAt;
+    }
+
+    public override DoAfterEvent Clone() => this;
+}
```
Comment (DEATHB4DEFEAT) [inline Content.Shared/Psionics/Events.cs:59]:
```suggestion
}

```

File: Content.Shared/Psionics/PsionicComponent.cs [hunk 8]
```diff
@@ -166,6 +167,18 @@ private set
         ///     unneccesary subs for unique psionic entities like e.g. Oracle.
         /// </summary>
         [DataField]
-        public List<String>? PsychognomicDescriptors = null;
+        public List<string>? PsychognomicDescriptors = null;
+
+        /// <summary>
+        ///     Used for tracking what spell a Psion is actively casting.
+        /// </summary>
```
Comment (DEATHB4DEFEAT) [inline Content.Shared/Psionics/PsionicComponent.cs:174]:
```suggestion
        /// Used for tracking what spell a Psion is actively casting
```

File: Content.Shared/Psionics/PsionicComponent.cs [hunk 9]
```diff
@@ -166,6 +167,18 @@ private set
...
+        [DataField]
+        public DoAfterId? DoAfter;
+
+        /// <summary>
+        ///     Popup to play if a Psion attempts to start casting a power while already casting one.
+        /// </summary>
```
Comment (DEATHB4DEFEAT) [inline Content.Shared/Psionics/PsionicComponent.cs:180]:
```suggestion
        /// Popup to play if a Psion attempts to start casting a power while already casting one
```

File: Resources/Locale/en-US/psionics/psionic-powers.ftl [hunk 10]
```diff
@@ -48,6 +49,27 @@ psionic-regeneration-power-initialization-feedback =
     I look within myself, finding a wellspring of life.
 psionic-regeneration-power-metapsionic-feedback = {CAPITALIZE($entity)} possesses an overwhelming will to live
 
+# Healing Word
+action-name-healing-word = Healing Word
+action-description-healing-word = Speak the Lesser Secret Of Life, and restore health to another.
+healing-word-power-description = Speak the Lesser Secret Of Life, and restore health to another.
+healing-word-power-initialization-feedback =
+    At the beginning of time, a word was spoken that brought life into the Spheres. Though it taxes my mind to know it,
+    this Secret is known to me now. I need only speak it.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/psionics/psionic-powers.ftl:58]:
```suggestion
    At the beginning of time, a word was spoken that brought life into the Spheres.
    Though it taxes my mind to know it, this Secret is known to me now.
    I need only speak it.
```

File: Resources/Locale/en-US/psionics/psionic-powers.ftl [hunk 11]
```diff
@@ -48,6 +49,27 @@ psionic-regeneration-power-initialization-feedback =
...
+healing-word-power-metapsionic-feedback = {CAPITALIZE($entity)} bears the Lesser Secret Of Life.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/psionics/psionic-powers.ftl:59]:
```suggestion
healing-word-power-metapsionic-feedback = {CAPITALIZE($entity)} bears the Lesser Secret of Life.
```

File: Resources/Locale/en-US/psionics/psionic-powers.ftl [hunk 12]
```diff
@@ -48,6 +49,27 @@ psionic-regeneration-power-initialization-feedback =
...
+healing-word-begin = {CAPITALIZE($entity)} mutters a word that brings both joy and pain alike to those who hear it.
+
+# Revivify
+action-name-revivify = Breath Of Life
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/psionics/psionic-powers.ftl:63]:
```suggestion
action-name-revivify = Breath of Life
```

File: Resources/Locale/en-US/psionics/psionic-powers.ftl [hunk 13]
```diff
@@ -48,6 +49,27 @@ psionic-regeneration-power-initialization-feedback =
...
+action-description-revivify = Speak the Greater Secret of Life, and restore another to life.
+revivify-power-description = Speak the Greater Secret of Life, and restore another to life.
+revivify-power-initialization-feedback =
+    For a moment, my soul journeys across time and space to the beginning of it all, there I hear it.
+    The Secret of Life in its fullness. I feel my entire existence burning out from within, merely by knowing it.
+    Power flows through me as a mighty river, begging to be released with a simple spoken word.
+revivify-power-metapsionic-feedback = {CAPITALIZE($entity)} bears the Greater Secret Of Life.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/psionics/psionic-powers.ftl:70]:
```suggestion
revivify-power-metapsionic-feedback = {CAPITALIZE($entity)} bears the Greater Secret of Life.
```

File: Resources/Prototypes/Actions/psionics.yml [hunk 14]
```diff
@@ -146,3 +146,76 @@
...
+    range: 6
+    itemIconStyle: BigAction
+    canTargetSelf: true
+    blacklist:
+      components:
+        - PsionicInsulation
+        - Mindbroken
+    event: !type:PsionicHealOtherPowerActionEvent
+            healingAmount:
+              groups: # These all get divided by the number of damage types in the group. So they're all -2.5.
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Actions/psionics.yml:168]: wuh

File: Resources/Prototypes/Actions/psionics.yml [hunk 15]
```diff
@@ -146,3 +146,76 @@
...
+                Burn: -60
+            rotReduction: 60
+            doRevive: true
+            powerName: Revivify
+            popupText: revivify-begin
+            playSound: true
+            minGlimmer: 10 # These also get multiplied by caster stats. So,
+            maxGlimmer: 15 # keeping in mind the ~3.5x multiplier, this spikes glimmer by as much as 60 points.
+            glimmerObviousSoundThreshold: 50
+            glimmerObviousPopupThreshold: 100
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Actions/psionics.yml:221]:
```suggestion
            glimmerObviousPopupThreshold: 100

```

File: Resources/Prototypes/Actions/psionics.yml [hunk 16]
```diff
@@ -146,3 +146,76 @@
...
+    blacklist:
+      components:
+        - PsionicInsulation
+        - Mindbroken
+    event: !type:PsionicHealOtherPowerActionEvent
+            healingAmount:
+              groups: # These all get divided by the number of damage types in the group. So they're all -15.
+                      # Additionally, they're multiplied by the caster's Amplification.
+                      # Which assuming this is the only power they have,
+                      # The multiplier is between 2.9 and 3.9
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Actions/psionics.yml:207]:
```suggestion
              # These all get divided by the number of damage types in the group. So they're all -15
              # Additionally, they're multiplied by the caster's Amplification, which,
              #   assuming this is the only power they have, the multiplier is between 2.9-3.9
              groups:
```

File: Resources/Prototypes/Actions/psionics.yml [hunk 17]
```diff
@@ -146,3 +146,76 @@
...
+                Genetic: -15
+                Toxin: -30
+                Airloss: -60 #Except airloss, which heals 30 per type.
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Actions/psionics.yml:210]:
```suggestion
                Airloss: -60 # Except airloss, which heals 30 per type
```

File: Resources/Prototypes/Psionics/psionics.yml [hunk 18]
```diff
@@ -140,3 +140,25 @@
...
+
+- type: psionicPower
+  id: RevivifyPower
+  name: Revivify
+  description: revivify-power-description
+  actions:
+    - ActionRevivify
+  initializationFeedback: revivify-power-initialization-feedback
+  metapsionicFeedback: revivify-power-feedback
+  amplificationModifier: 2.5 # An extremely rare and dangerous power.
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Psionics/psionics.yml:163]:
```suggestion
  amplificationModifier: 2.5 # An extremely rare and dangerous power
```

File: Resources/Prototypes/Roles/Jobs/Civilian/chaplain.yml [hunk 19]
```diff
@@ -5,8 +5,8 @@
   playTimeTracker: JobChaplain
   requirements:
     - !type:CharacterDepartmentTimeRequirement
-      department: Epistemics # DeltaV - Epistemics Department replacing Science
-      min: 14400 #DeltaV 4 hours
+      department: Epistemics # Chaplain is now one of the station's "Crew Aligned Wizards"
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Roles/Jobs/Civilian/chaplain.yml:8]:
```suggestion
      department: Epistemics # Chaplain is now one of the station's "Crew-Aligned Wizards"
```

File: Resources/Textures/Interface/VerbIcons/healing_word.png
Comment (DEATHB4DEFEAT) [inline Resources/Textures/Interface/VerbIcons/healing_word.png:1]: Put these in an RSI somewhere else, this folder is a mess and that attribution.txt is horrible to see when we have [RGA](https://docs.spacestation14.com/en/specifications/robust-generic-attribution.html)

**Review by DEATHB4DEFEAT — DISMISSED** [review]

File: Resources/Textures/Interface/VerbIcons/healing_word.png
Comment (VMSolidus) [reply Resources/Textures/Interface/VerbIcons/healing_word.png:1]: Done

**Review by Peptide90 — APPROVED** [review]

**Pull request merged**